### PR TITLE
Add commit hash generation to assemblies

### PR DESCRIPTION
**Pull request by attilah**
The git hash logic variable set logic was taken from: https://github.com/dotnet/roslyn/blob/master/build/Targets/Versions.props#L56

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -116,4 +116,40 @@
   <PropertyGroup>
     <PackageOutputPath Condition=" '$(PackageOutputPath)'=='' ">$(SourceRoot)/Artifacts/$(Configuration)</PackageOutputPath>
   </PropertyGroup>
+
+  <Choose>
+    <When Condition="'$(OfficialBuild)' != 'true'">
```
Comment (jdom) [inline Directory.Build.props]: VSO should define this?

File: Directory.Build.props [hunk 2]
```diff
@@ -116,4 +116,40 @@
...
+      </PropertyGroup>
+    </When>
+    <When Condition="'$(BUILD_SOURCEVERSION)' != ''">
+      <PropertyGroup>
+        <GitHeadSha>$(BUILD_SOURCEVERSION)</GitHeadSha>
+      </PropertyGroup>
+    </When>
+    <When Condition="'$(BUILD_SOURCEVERSION)' == '' AND '$(GIT_COMMIT)' != ''">
+      <PropertyGroup>
+        <GitHeadSha>$(GIT_COMMIT)</GitHeadSha>
```
Comment (jdom) [inline Directory.Build.props]: what is this? Jenkins?

File: Directory.Build.props [hunk 3]
```diff
@@ -116,4 +116,40 @@
...
+        <HeadFileContent Condition="Exists('$(DotGitDir)/HEAD')">$([System.IO.File]::ReadAllText('$(DotGitDir)/HEAD').Trim())</HeadFileContent>
+        <RefPath Condition="$(HeadFileContent.StartsWith('ref: '))">$(DotGitDir)/$(HeadFileContent.Substring(5))</RefPath>
+        <GitHeadSha Condition="'$(RefPath)' != '' AND Exists('$(RefPath)')">$([System.IO.File]::ReadAllText('$(RefPath)').Trim())</GitHeadSha>
+        <GitHeadSha Condition="'$(HeadFileContent)' != '' AND '$(RefPath)' == ''">$(HeadFileContent)</GitHeadSha>
+      </PropertyGroup>
+    </Otherwise>
+  </Choose>
+
+  <PropertyGroup>
+    <InformationalVersion>$(VersionPrefix)-$(VersionSuffix). Commit Hash: $(GitHeadSha)</InformationalVersion>
```
Comment (jdom) [inline Directory.Build.props]: I think you should use `Version`... VersionPrefix and Suffix **may** be used to define the version, but not necessarily (and we are no longer using them)

File: Directory.Build.props [hunk 4]
```diff
@@ -116,4 +116,40 @@
   <PropertyGroup>
     <PackageOutputPath Condition=" '$(PackageOutputPath)'=='' ">$(SourceRoot)/Artifacts/$(Configuration)</PackageOutputPath>
   </PropertyGroup>
+
+  <Choose>
```
Comment (jdom) [inline Directory.Build.props]: condition if GitHeadSha wasn't already defined?

Comment (attilah) [reply Directory.Build.props]: I'd put an animgif here "simply no one does define GitHeadSha" :D

File: Directory.Build.props (hunk 1 above)
Comment (attilah) [reply Directory.Build.props]: Yes

File: Directory.Build.props (hunk 2 above)
Comment (attilah) [reply Directory.Build.props]: Yup :)

File: Directory.Build.targets [hunk 5]
```diff
@@ -17,4 +17,9 @@
   </PropertyGroup>
   <Import Project="$(SourceRoot)src/Orleans.CodeGeneration.Build/build/Microsoft.Orleans.OrleansCodeGenerator.Build.targets" Condition=" '$(OrleansBuildTimeCodeGen)' == 'true' " />
   <!--End Orleans -->
+
+  <!-- Set InformationVersion here, since $(Version) is already set at this point. -->
+  <PropertyGroup>
+    <InformationalVersion>$(Version). Commit Hash: $(GitHeadSha)</InformationalVersion>
```
Comment (jdom) [inline Directory.Build.targets]: why is this in targets and not props file?

Comment (jdom) [reply Directory.Build.targets]: Ah, it's because this runs after the Version variable is set? I just read the comment above :)

**Pull request merged**